src/priceHistory/priceHistoryChart.tsx

```diff
@@ -0,0 +1,357 @@
+  )
+  const localCurrencyCode = useSelector(getLocalCurrencyCode)
+  const localExchangeRate = useSelector(usdToLocalCurrencyRateSelector)
+  const dispatch = useDispatch()
+
+  const dollarsToLocal = useCallback(
+    (amount: BigNumber.Value | null) =>
+      convertDollarsToLocalAmount(amount, localCurrencyCode ? localExchangeRate : 1),
+    [localExchangeRate]
+  )
+  const displayLocalCurrency = useCallback(
+    (amount: BigNumber.Value) =>
+      getLocalCurrencyDisplayValue(amount, localCurrencyCode || LocalCurrencyCode.USD, true),
+    [localCurrencyCode]
+  )
+
+  useEffect(() => {
+    const latestTimestamp = priceHistoryPrices.at(-1)?.priceFetchedAt ?? 0
+    if (priceHistoryPrices.length > 0 && latestTimestamp > Date.now() - ONE_HOUR_IN_MILLIS) {
+      return
+    }
+
+    dispatch(
+      fetchPriceHistoryStart({
+        tokenId,
+        startTimestamp: Date.now() - ONE_DAY_IN_MILLIS * 30,
+        endTimestamp: Date.now(),
+      })
+    )
+  }, [])
+
+  if (priceHistoryStatus === 'loading' && priceHistoryPrices.length === 0) {
+    return <Loader />
+  } else if (priceHistoryPrices.length === 0) {
+    return null
+  }
+
+  // Create chart data from price history
+  const chartData = []
+  let lastTimestampAdded, highPrice, lowPrice
+  for (let i = 0; i < priceHistoryPrices.length; i++) {
+    // Check if price is the highest or lowest price and if so store it
+    if (!highPrice || priceHistoryPrices[i].priceUsd > highPrice.priceUsd) {
+      highPrice = priceHistoryPrices[i]
+    }
+    if (!lowPrice || priceHistoryPrices[i].priceUsd < lowPrice.priceUsd) {
+      lowPrice = priceHistoryPrices[i]
+    }
+
+    // Only grab one price per chart step & the most recent price
+    if (
+      lastTimestampAdded &&
+      priceHistoryPrices[i].priceFetchedAt - lastTimestampAdded <
+        ONE_HOUR_IN_MILLIS * CHART_STEP_IN_HOURS &&
+      i !== priceHistoryPrices.length - 1
+    ) {
+      continue
+    } else {
+      lastTimestampAdded = priceHistoryPrices[i].priceFetchedAt
+      const { priceUsd } = priceHistoryPrices[i]
+      const localAmount = dollarsToLocal(priceUsd)
+      chartData.push({
+        amount: localAmount ? localAmount.toNumber() : 0,
+        displayValue: localAmount ? displayLocalCurrency(localAmount) : '',
+        priceFetchedAt: priceHistoryPrices[i].priceFetchedAt,
+      })
+    }
+  }
+
+  // Make sure the highest and lowest prices are included in the chart
+  if (highPrice) {
+    chartData.push({
+      amount: dollarsToLocal(highPrice.priceUsd)?.toNumber() || 0,
+      displayValue: displayLocalCurrency(dollarsToLocal(highPrice.priceUsd) || 0),
+      priceFetchedAt: highPrice.priceFetchedAt,
+    })
+  }
+  if (lowPrice) {
+    chartData.push({
+      amount: dollarsToLocal(lowPrice.priceUsd)?.toNumber() || 0,
+      displayValue: displayLocalCurrency(dollarsToLocal(lowPrice.priceUsd) || 0),
+      priceFetchedAt: lowPrice.priceFetchedAt,
+    })
+  }
```

Comment:
is it strictly necessary to include these? I guess these could cause uneven data points and I assume the sorting and unique steps below are only required because of adding these..

Reply:
With the filtering based on step CHART_STEP_IN_HOURS it's very possible if we have a lot of data points we will exclude the highest and lowest prices. The sorting is needed to get this into the right place, but I didn't notice much of a performance hit sorting the filtered list. Displaying the unfiltered list was a much bigger bottleneck.